parser.y

```diff
@@ -422,11 +439,93 @@ init_declarator: declarator { $$ = $1; }
 /* 6.7.2 Type specifiers */
 /* TODO: support multiple data types */
 type_specifier: INT { $$ = std::make_unique<PrimType>(PrimitiveType::kInt); }
-  /* TODO: struct or union specifier */
+  | struct_or_union_specifier { $$ = $1; }
   /* TODO: enum specifier */
   /* TODO: typedef name */
   ;
 
+struct_or_union_specifier: struct_or_union id_opt LEFT_CURLY struct_declaration_list RIGHT_CURLY {
+    // Field types for variable 'type' are unknown until now.
+    auto type = $1;
+    auto decl_id = $2;
+    auto field_list = $4;
+    auto field_types = std::vector<std::unique_ptr<Type>>{};
+    for (const auto& field : field_list) {
+      field_types.push_back(field->type->Clone());
+    }
+
+    if (dynamic_cast<StructType*>(type.get())) {
+      type = std::make_unique<StructType>(std::move(field_types));
+    } else {
+      type = std::make_unique<UnionType>(std::move(field_types));
+    }
+
+    if (decl_id) {
+      $$ = std::make_unique<RecordDeclNode>(Loc(@2), std::move(decl_id->id), std::move(type), std::move(field_list));
+    } else {
+      $$ = std::make_unique<RecordDeclNode>(Loc(@2), /* no id */ "", std::move(type), std::move(field_list));
+    }
```

Comment:
Since the two condition bodies are only different in the `id`, I think we can use the ternary operator (`?:`).
```suggestion
    $$ = std::make_unique<RecordDeclNode>(Loc(@2), decl_id ? std::move(decl_id->id) : "", std::move(type), std::move(field_list));
```